cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,491 @@
+
+    Creates the following attributes (see parent constructor for more):
+    - cpo_model: object, CP Optimizers model object
+    """
+
+    @staticmethod
+    def supported():
+        return CPM_cpo.installed() and CPM_cpo.executable_installed()
+
+    @staticmethod
+    def installed():
+        # try to import the package
+        try:
+            import docplex.cp as docp
+            return True
+        except ModuleNotFoundError:
+            return False
+
+    @staticmethod
+    def executable_installed():
+        if shutil.which('cpoptimizer') is None:
+            return False
+        else:
+            return True
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: str, name of a subsolver (optional)
+        """
+        if not self.installed():
+            raise Exception("CPM_cpo: Install the python package 'docplex'")
+
+        if not self.executable_installed():
+            raise Exception("You need to install the CPLEX Optimization Studio to use this solver.")
+
+        import docplex.cp.model as dom
+        assert subsolver is None
+
+        # initialise the native solver object
+        # [GUIDELINE] we commonly use 3-letter abbrivations to refer to native objects:
+        #           OR-tools uses ort_solver, Gurobi grb_solver, Exact xct_solver...
+        self.cpo_model = dom.CpoModel()
+
+        # initialise everything else and post the constraints/objective
+        # [GUIDELINE] this superclass call should happen AFTER all solver-native objects are created.
+        #           internally, the constructor relies on __add__ which uses the above solver native object(s)
+        super().__init__(name="cpo", cpm_model=cpm_model)
+
+
+    def solve(self, time_limit=None, **kwargs):
+        """
+            Call the CP Optimizer solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            # [GUIDELINE] Please document key solver arguments that the user might wish to change
+            #       for example: assumptions=[x,y,z], log_output=True, var_ordering=3, num_cores=8, ...
+            # [GUIDELINE] Add link to documentation of all solver parameters
+        """
+
+        # [GUIDELINE] if your solver supports solving under assumptions, add `assumptions` as argument in header
+        #       e.g., def solve(self, time_limit=None, assumptions=None, **kwargs):
+        #       then translate assumptions here; assumptions are a list of Boolean variables or NegBoolViews
+
+        # call the solver, with parameters
+        self.cpo_result = self.cpo_model.solve(LogVerbosity='Quiet', TimeLimit=time_limit, **kwargs)
```

Comment:
by setting LogVerbosity here it can't be overwritten potentially? If you add LogVerbosity='Quiet' to our def solve() then perhaps it would?

Reply:
Ah yes, should check if it's in the kwargs first!